FOCUS #169: Pricing Unit Dimension (#179)
FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec: specification/dimensions/pricingunit.md

# Pricing Unit

> The Pricing Unit is a measurement unit specified by the provider for determining unit prices associated with the usage or the purchase of a specific resource and/or service. Common examples include the number of hours a compute appliance was running for a given period, the number of gigabytes * the number of hours (gigabyte-hours) for a storage appliance, or a cumulative count of requests for a network appliance or API service. For certain services providers can choose block pricing i.e., million requests as a base unit. PricingUnit is complementary to and describes how [QuantityInPricingUnit](#quantityinpricingunit) is measured. While PricingUnit may appear to relate to resource and/or service consumption, PricingUnit is primarily focused on pricing and cost. It is essential not to confuse PricingUnit with [UsageUnit](#usageunit), which is dedicated to tracking resource and/or service consumption and is often at a lower granularity than PricingUnit.

Reviewer: ```suggestion
The Pricing Unit is a measurement unit specified by the provider for determining unit prices associated with the usage or the purchase of a specific resource and/or service. Common examples include the number of hours a compute appliance was running for a given period, the number of gigabytes * the number of hours (gigabyte-hours) for a storage appliance, or a cumulative count of requests for a network appliance or API service. For certain services providers can choose block pricing i.e., million requests as a base unit. PricingUnit is complementary to and describes how [QuantityInPricingUnit](#quantityinpricingunit) is measured. The PricingUnit is primarily focused on pricing and cost. It should not be confused with [UsageUnit](#usageunit), which is dedicated to tracking resource and/or service consumption and is often at a lower granularity than the PricingUnit.
```

Reviewer: moving suggestion to another conversation which encapsulates additional suggestions for improvements

> The PricingUnit column MUST be present in the billing data. This column MUST be of type String. It MUST NOT contain null if [QuantityInPricingUnit](#quantityinpricingunit) is not null. The PricingUnit value MUST match what is shown on the invoice if a pricing unit is present on the invoice.

Reviewer: ```suggestion
The PricingUnit column MUST be present in the billing data. This column MUST be of type String. It MUST NOT contain null if [QuantityInPricingUnit](#quantityinpricingunit) is not null. The PricingUnit MUST match the corresponding value published in the price list, and if present on the invoice, it MUST also match the value on the invoice.
```

> The PricingUnit column MUST be present in the billing data. This column MUST be of type String. It MUST NOT contain null if [QuantityInPricingUnit](#quantityinpricingunit) is not null. The PricingUnit value MUST match what is shown on the invoice if a pricing unit is present on the invoice.

Reviewer: ```suggestion
The PricingUnit column MUST be present in the billing data. This column MUST be of type String. It MUST NOT contain null if [QuantityInPricingUnit](#quantityinpricingunit) is not null. The PricingUnit MUST match the corresponding value published in the price list. If the level of detail provided on invoice includes pricing units, PricingUnit values must be consistent in both the billing data and the invoice for the same billing period.
```

# Pricing Unit

> The Pricing Unit represents a provider specified pricing measurement unit, which forms the foundation for determining unit prices of utilizing a particular resource and/or service. Common examples include the number of hours a compute appliance was running for a given period, the number of gigabytes * the number of hours (gigabyte-hours) for a storage appliance, or a cumulative count of requests for a network appliance or API service. For certain services providers can choose block pricing i.e., million requests as a base unit.

Reviewer 2: ```suggestion
The Pricing Unit is a measurement unit specified by the provider, serving as the basis for calculating unit prices associated with using a specific resource and/or service. Examples include the duration a compute appliance is active, the product of gigabytes and hours (gigabyte-hours) for a storage appliance, or the total count of requests for a network appliance or API service. In some cases, providers may opt for block pricing, such as pricing per million requests as the pricing unit. The Pricing Unit, in conjunction with [Quantity In Pricing Unit](#quantityinpricingunit), defines and describes the measurement method for determining costs.
```

Reviewer 2: My suggestion only.. use all/parts if it makes sense to do so.

Author: I like this.  I'm wondering how serious @flanakin is about the word "basis" vs "foundation", but I'm inclined to accept this as is and then we can discuss that part.  Thoughts?

Reviewer 3: Sounds very robust, the wording is clear and the main message stays the same 👍

Reviewer: I'd like to challenge the term `calculating` :). I feel like `determining` is a better fit in this context.

Reviewer 3: @ijurica - you are referring to the fact that `Pricing Unit` is more of an information than a `p * q = c` dimension?

Reviewer: Costs are usually calculated and unit prices determined/specified therefore, in this particular sentence, I would prefer the term **determining** over **calculating**. 

_The Pricing Unit is a measurement unit specified by the provider, serving as the basis for **calculating** unit prices associated with using a specific resource and/or service._

Reviewer 3: Aligned during the call that we should determining 👍

PricingUnit is complementary to and describes how [QuantityInPricingUnit](#quantityinpricingunit) is measured.

> 

Reviewer 3: ```suggestion
The Pricing Unit for a particular line item may differ from the published Usage Unit (and complementary Quantity in Usage Unit) when providers use different units, unit increments, or columns for calculating cost. 
```

Reviewer: I've added a remark regarding the UsageUnit/PricingUnit within Usage Unit PR as well. This seems a bit unclear and implies that there are UCs in which providers use different columns for cost calculation?!

Suggestions:
- focus on differing purposes (primary distinction?!) 
- avoid mentioning that the values may differ and when that may be the case as covering those UCs seams a bit challenging

We might consider something like this: 
_While PricingUnit and QuantityInPricing may appear to relate to resource and/or service consumption, their primary focus is on pricing and cost, so it's essential not to confuse them with UsageUnit and QuantityInUsageUnit, which are dedicated to tracking actual resource and service consumption._

Rephrase? Simplify? Definitely!

Reviewer: Discussed at a meeting and agreed to emphasize the difference in purpose rather than mentioning that UsageUnit and PricingUnit may differ and in which cases. A similar note or remark will be included in the definition of all four columns - PricingUnit, QuantityInPricingUnit, UsageUnit, QuantityInPricingUnit.

> The PricingUnit column MUST be present in the billing data. This column MUST be of type String. It MUST NOT contain null if [QuantityInPricingUnit](#quantityinpricingunit) is not null. The PricingUnit value MUST match what is shown on the invoice if a pricing unit is present on the invoice.

Reviewer 2: Is there not a standard set of units (or a combination) we're recommending like the usage unit team is doing? 

cc @amitw-g @ijurica

Reviewer: I believe UsageQuantity + UsageUnit pair can also be described as a semi-normalized version of QuantityinPricingUnit + PricingUnit (not-normalized) pair.  I agree, we should mention that in the supporting content. As for the definition, phrasing this is a challenge :(.  I suggest we get back to that at a later stage, once names and definitions of UsageUnit and Quantity are finalized.

Few more notes about the PricingUnit...

**For AWS, Microsoft, GCP:**

PricingUnit comes down to mapped/renamed Provider's column

**For OCI:**

There is a column in billing data which we could use but it might be better to fetch the corresponding value from Public Pricing API (we already plan to use it for ListUnitPrice). 
Example:

- values in billing data: `ONE GiB HOURS STORAGE_SIZE`, `ONE GiB HOURS MEMORY`
- corresponding value in Public Pricing API: `Gigabyte Per Hour`

Reviewer 3: We have discussed that it may be very hard to ship 1.0 with standard set of units, my believe is that we should keep examples as in the definition to get the idea of what is this column referring to. Later on we will see what is the value of those being standardised

Reviewer 4: We have a set of recommended units, but the UsageUnit was transitioned from a "MUST use Allowed Values" to "SHOULD use recommended values" - maybe we should move that part into an appendix and refer to it from both?

Reviewer: - notes regarding non-normalized vs normalized and future plans provided within supporting content
- inclusion of PricingUnit - UsageUnit cross-references in progress
- current values observed in billing data for OCI provided

> The PricingUnit column MUST be present in the billing data. This column MUST be of type String. It MUST NOT contain null if [QuantityInPricingUnit](#quantityinpricingunit) is not null. The PricingUnit value MUST match what is shown on the invoice if a pricing unit is present on the invoice.

Reviewer 2: Should we mention the difference between usage unit and pricing? Im not suggesting we do but want to make sure that was discussed and agreed on.

Reviewer 5: Would that be an appendix? (Asking not stating 🙂)

Reviewer 3: TL;DR
If we decide to keep comments regarding PricingUnit in UsageUnit spec then we should do the other way around here as well.

I think as PricingUnit is complementary we literally had to put this information there. For considerations and relations to non-complementary units we may want to stick to appendix BUT as the UsageUnit contributors are commenting on PricingUnit we can add those comments.

## Column ID

PricingUnit

## Display name

Pricing Unit

## Description

> Represents a provider specified pricing measurement unit, which forms the foundation for determining unit prices of utilizing a particular resource and/or service.

Reviewer 2: ```suggestion
A measurement unit specified by the provider, serving as the basis for calculating unit prices associated with using a specific resource and/or service.
```

Reviewer 5: I would love to avoid the term "basis". It's jargon that most people don't understand based on my conversations with customers outside of Finance.

Reviewer 3: As a non-finance+non-native I would say that _basis_ is as understandable as _foundation_ in this context. 
@flanakin / @udam-f2 - maybe we should ship version 1.0 and eventually rethink such wordings after wider community feedback?

Reviewer 3: @udam-f2 I had a quick chat with @flanakin and Michael mentions that basis can be widely unknown in a non-finance world based on his past experience.
What do you think about such small tweak?
**_serving as the foundation (aka basis) for calculating unit prices_**

Reviewer 2: I'm ok with foundation.

> # Pricing Unit

Reviewer 5: There have been conversations about having a semi/normalized pricing unit post-1.0 with a separate block size. If we do that, I'm assuming we'll want to keep PricingUnit and UsageUnit aligned in naming and values. Can we rename this to `PricingMeasurement` to keep the provider-specific value separate?

Reviewer: Discussed at the meeting and decided to rename the **PricingUnit** column to **PricingMeasure** (non-normalized).

> The Pricing Unit is a measurement unit specified by the provider for determining unit prices associated with the usage or the purchase of a specific resource and/or service. Common examples include the number of hours a compute appliance was running for a given period, the number of gigabytes * the number of hours (gigabyte-hours) for a storage appliance, or a cumulative count of requests for a network appliance or API service. For certain services providers can choose block pricing i.e., million requests as a base unit. Pricing Unit is complementary to and describes how [Quantity In Pricing Unit](#quantityinpricingunit) is measured. The Pricing Unit is primarily focused on pricing and cost and it should not be confused with the [Usage Unit](#usageunit), which is dedicated to tracking resource and/or service consumption and is often at a lower granularity than the Pricing Unit.

Reviewer: - Improve the PricingUnit and UsageUnit definitions, especially the PricingUnit - UsageUnit cross-references by incorporating the following suggestions: 
  - _The Pricing Unit indicates how a provider rates the measured usage and purchase quantities after pricing rules are applied. It may differ from the physical measurement unit or Usage Unit due to pricing rules, such as block pricing. The Pricing Unit is commonly used for scenarios like verifying unit prices against the price sheet and should not be directly correlated with the physical resource or service usage._
  - _The Usage Unit indicates how service usage and purchases are measured by the provider. The Usage Unit is a standard measure of physical quantity before the pricing has been applied. The Usage Unit describes and is used in conjunction with Quantity In Usage Unit to gain an understanding of how much of a given product or service was used or purchased. Usage Unit may be a lower-level granularity than Pricing Unit, which applies provider-specific pricing rules. The Usage Unit is commonly used for scenarios like verifying how much of a resource or service was used with external metrics and logging systems to optimize usage patterns and should not be used to correlate directly to the unit price._
- Ensure that the definitions of related columns (PricingUnit, QuantityInPricingUnit, UsageUnit, and QuantityInUsageUnit) are aligned.
- Discuss these changes with both VS1 and VS3 team members.

Reviewer 3: First 2 points covered, discussion was taken and maintainers reviews should cover the "discussion" part

> The Pricing Unit is a measurement unit specified by the provider for determining unit prices associated with the usage or the purchase of a specific resource and/or service. Common examples include the number of hours a compute appliance was running for a given period, the number of gigabytes * the number of hours (gigabyte-hours) for a storage appliance, or a cumulative count of requests for a network appliance or API service. For certain services providers can choose block pricing i.e., million requests as a base unit. Pricing Unit is complementary to and describes how [Quantity In Pricing Unit](#quantityinpricingunit) is measured. The Pricing Unit is primarily focused on pricing and cost and it should not be confused with the [Usage Unit](#usageunit), which is dedicated to tracking resource and/or service consumption and is often at a lower granularity than the Pricing Unit.

Reviewer 2: ```suggestion
The Pricing Unit is the provider-specified measurement unit for determining the unit prices associated with the use or purchase of a specific resource or service. Examples include measuring the hours a compute appliance runs, calculating gigabyte-hours for storage appliances, or tallying requests for network appliances or API services. In some cases, providers may opt for block pricing, such as using a million requests as a foundational unit. The Pricing Unit complements and describes how the measurement of [Quantity In Pricing Unit](#quantityinpricingunit) is conducted. The Pricing Unit pertains to pricing and cost determination and should not be confused with the [Usage Unit](#usageunit), which tracks resource and service consumption, typically at a finer granularity than the Pricing Unit.
```

Reviewer 2: Use if more concise. Credit ChatGPT :)

Reviewer 3: We took some bits of those and commited pre-maintainer version to the repo

> The Pricing Unit is a measurement unit specified by the provider for determining unit prices associated with the usage or the purchase of a specific resource and/or service. Common examples include the number of hours a compute appliance was running for a given period, the number of gigabytes * the number of hours (gigabyte-hours) for a storage appliance, or a cumulative count of requests for a network appliance or API service. For certain services providers can choose block pricing i.e., million requests as a base unit. Pricing Unit is complementary to and describes how [Quantity In Pricing Unit](#quantityinpricingunit) is measured. The Pricing Unit is primarily focused on pricing and cost and it should not be confused with the [Usage Unit](#usageunit), which is dedicated to tracking resource and/or service consumption and is often at a lower granularity than the Pricing Unit.

Reviewer: Taking into account the original version and the suggestions, this is what I'm currently voting for

> The Pricing Unit is the provider-specified measurement unit for determining the unit prices associated with the use or purchase of a specific resource and/or service. The Pricing Unit indicates how a provider rates the measured usage and purchase quantities after pricing rules, such as block pricing, are applied. Common examples include the number of hours for compute appliance runtime, gigabyte-hours for a storage appliance, or an accumulated count of requests for a network appliance or API service. In some cases, providers may opt for block pricing, such as using a million requests as the Pricing Unit instead of pricing individual requests. The Pricing Unit complements and describes how the measurement of [Quantity In Pricing Unit](#quantityinpricingunit) is conducted. The Pricing Unit pertains to pricing and cost determination and should not be confused with the [Usage Unit](#usageunit), which tracks resource and service consumption, typically at a finer granularity than the Pricing Unit.

Reviewer 2: Looking to see if there's a way to make the description more concise. The second sentence seems unnecessary since we describe the block pricing provision a little further down anyway. 

```suggestion
The Pricing Unit is the provider-specified measurement unit for determining the unit prices associated with the use or purchase of a specific resource and/or service. Common examples include the number of hours for compute appliance runtime, gigabyte-hours for a storage appliance, or an accumulated count of requests for a network appliance or API service. In some cases, providers may use block pricing, like a million requests as the Pricing Unit. Pricing Unit complements the [Quantity In Pricing Unit](#quantityinpricingunit) measurement. Distinct from the [Usage Unit](#usageunit), it focuses on pricing, not resource and service consumption, often at a coarser granularity.
```

Reviewer: Addressed in [a subsequent conversation](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/179#discussion_r1382309783)

> The Pricing Unit is the provider-specified measurement unit for determining the unit prices associated with the use or purchase of a specific resource and/or service. The Pricing Unit indicates how a provider rates the measured usage and purchase quantities after pricing rules, such as block pricing, are applied. Common examples include the number of hours for compute appliance runtime, gigabyte-hours for a storage appliance, or an accumulated count of requests for a network appliance or API service. In some cases, providers may opt for block pricing, such as using a million requests as the Pricing Unit instead of pricing individual requests. The Pricing Unit complements and describes how the measurement of [Quantity In Pricing Unit](#quantityinpricingunit) is conducted. The Pricing Unit pertains to pricing and cost determination and should not be confused with the [Usage Unit](#usageunit), which tracks resource and service consumption, typically at a finer granularity than the Pricing Unit.

Reviewer: ```suggestion
The Pricing Unit is the provider-specified measurement unit for determining unit prices, indicating how a provider rates measured usage and purchase quantities after applying pricing rules like block pricing. Common examples include the number of hours for compute appliance runtime, gigabyte-hours for a storage appliance, or an accumulated count of requests for a network appliance or API service. Pricing Unit complements the [Quantity In Pricing Unit](#quantityinpricingunit) measurement. Distinct from the [Usage Unit](#usageunit), it focuses on pricing, not resource and service consumption, often at a coarser granularity.
```

Reviewer: Similar to the [previous suggestion](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/179#discussion_r1382305221). The main/only difference: referring to block pricing as a pricing rule rather than mentioning it via an example.

Reviewer: introduced a minor modification to align with the definition of QuantityInPricingUnit

> # Pricing Unit

Reviewer 5: I really like the alignment to Unit Format for consistency. My only request would be to also add PricingBlockSize as the unit without the block size isn't enough information to understand how pricing works.

Reviewer 2: Block size can be a fast follow in v1.x since it's only necessary in a small number of (still important) use cases. Added here: [FOCUS Data Model](https://docs.google.com/spreadsheets/d/148Hdg1ws1Co3MDI65T_KtGCAzbtmFbzOU0pGvRnmbq8/edit#gid=1318248972) as 1.x

Reviewer: One of the notes in the Discussion/Scratch space refers to some suggestions for 1.0+

> The Pricing Unit represents a provider-specified measurement unit for determining unit prices, indicating how a provider rates measured usage and purchase quantities considering pricing rules like block pricing. Common examples include the number of hours for compute appliance runtime (e.g. `Hours`), gigabyte-hours for a storage appliance (e.g., `GB-Hours`), or an accumulated count of requests for a network appliance or API service (e.g., `1000 Requests`). Pricing Unit complements the [Pricing Quantity](#pricingquantity) metric. Distinct from the [Usage Unit](#usageunit), it focuses on pricing and cost, not resource and service consumption, often at a coarser granularity.

Reviewer 5: This uses an example with a block size, but I thought the Unit Format doesn't not want numbers in the value, right? Should we remove that example?

Reviewer 6: Is a quantifier dimension eg `pricingBlockSize` in 1.0 or post-1.0?

Reviewer 2: Pricing Unit is logically the same as what providers have today with a recommendation to use consistent formatting specified in Unit Format. Block size can be a fast follow in v1.0+ since it's only necessary in a small number of (still important) use cases. Primary need is for providing a way for users to make sense of the rates/quantities/cost and that's accomplished by this column.

Reviewer: One of the notes in the Discussion/Scratch space refers to some suggestions for 1.0+

> The Pricing Unit represents a provider-specified measurement unit for determining unit prices, indicating how a provider rates measured usage and purchase quantities after applying pricing rules like block pricing. Common examples include the number of hours for compute appliance runtime (e.g. `Hours`), gigabyte-hours for a storage appliance (e.g., `GB-Hours`), or an accumulated count of requests for a network appliance or API service (e.g., `1000 Requests`). Pricing Unit complements the [Pricing Quantity](#pricingquantity) metric. Distinct from the [Usage Unit](#usageunit), it focuses on pricing and cost, not resource and service consumption, often at a coarser granularity.

Reviewer: Already taken care of
